### PR TITLE
Feat/aut 553/remote publish class content

### PR DESCRIPTION
**Pull request by bartlomiejmarszal**
This feature allow user to publish all class instances to chosen remote environments. 

https://oat-sa.atlassian.net/browse/AUT-553

### CODE REVIEW

**Review by bziondik — APPROVED** [review]
FE part:
- [ ]  New code is covered by tests (if applicable)
- [ ]  Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x]  New code is respecting code style rules
- [x]  New code is respecting best practices
- [x]  New code is not subject to concurrency issues (if applicable)
- [x]  Feature is working correctly on my local machine (if applicable)
- [x]  Acceptance criteria are respected
- [x]  Pull request title and description are meaningful

File: model/publishing/environment/EnvironmentResourceValidatorTrait.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+            )
+        );
+
+        if ($publishingEnabled === false) {
+            throw new PublishingInvalidArgumentException(
+                __(sprintf('Remote publishing is disabled for environment with URI "%s"', $environment->getUri()))
+            );
+        }
+    }
+}
```
Comment (mike-ionut-mihai-sandu-tao) [inline model/publishing/environment/EnvironmentResourceValidatorTrait.php:51]: let's format with PSR and add /n

File: model/publishing/environment/EnvironmentResourceValidatorTrait.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+declare(strict_types=1);
+
+namespace oat\taoPublishing\model\publishing\environment;
+
+use core_kernel_classes_Resource;
+use oat\taoPublishing\model\PlatformService;
+use oat\taoPublishing\model\publishing\exception\PublishingInvalidArgumentException;
+
+trait EnvironmentResourceValidatorTrait
```
Comment (mike-ionut-mihai-sandu-tao) [inline model/publishing/environment/EnvironmentResourceValidatorTrait.php:29]: should we test this ?

Comment (mike-ionut-mihai-sandu-tao) [reply model/publishing/environment/EnvironmentResourceValidatorTrait.php:29]: as i see you just extracted , but let's at least add a technical debt  ticket or similar

File: model/publishing/delivery/PublishingClassDeliveryService.php [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    use ServiceLocatorAwareTrait;
+    use OntologyAwareTrait;
+    use EnvironmentResourceValidatorTrait;
+
+    /**
+     * @return TaskInterface[]
+     * @throws PublishingInvalidArgumentException
+     * @throws core_kernel_persistence_Exception
+     */
+    public function publish(core_kernel_classes_Class $class, array $environments): array
```
Comment (mike-ionut-mihai-sandu-tao) [inline model/publishing/delivery/PublishingClassDeliveryService.php:49]: test by mocking the queue ?

**Review by mike-ionut-mihai-sandu-tao — APPROVED** [review]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful

**Review by bziondik — APPROVED** [review]

**Review by Silvia-loza — APPROVED** [review]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [X] New code is respecting code style rules
- [X] New code is respecting best practices
- [X] New code is not subject to concurrency issues (if applicable)
- [X] Feature is working correctly on my local machine (if applicable)
- [X] Acceptance criteria are respected
- [X] Pull request title and description are meaningful

File: model/publishing/delivery/RemotePublishingService.php [hunk 4]
```diff
@@ -112,7 +113,7 @@ private function collectAndValidateEnvironmentResources(array $environments): ar
         $results = [];
         foreach ($environments as $environmentUri) {
             $environmentResource = $this->getResource($environmentUri);
-            $this->validateRemoteEnvironment($environmentResource);
+            $this->validateEnvironment($environmentResource);
```
Comment (pnal) [inline model/publishing/delivery/RemotePublishingService.php:116]: this method throws PublishingInvalidArgumentException please update phpDoc

**Review by mike-ionut-mihai-sandu-tao — APPROVED** [review]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [x] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful

File: model/publishing/delivery/PublishingClassDeliveryService.php [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    use OntologyAwareTrait;
+    use EnvironmentResourceValidatorTrait;
+
+    public const SERVICE_ID = 'taoPublishing/PublishingClassDeliveryService';
+    public const OPTION_MAX_RESOURCE = 'maxResource';
+
+    /**
+     * @return TaskInterface[]
+     * @throws PublishingInvalidArgumentException
+     * @throws core_kernel_persistence_Exception
```
Comment (pnal) [inline model/publishing/delivery/PublishingClassDeliveryService.php:51]:
```suggestion
     * @throws core_kernel_persistence_Exception
     * @throws ExceedingMaxResourceAmountException
```

File: test/unit/model/publishing/delivery/PublishingClassDeliveryServiceTest.php [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+                    [
+                        'subject' => 'someResourceUri4'
+                    ],
+                ]
+            );
+
+        $this->subject->publish($this->classMock, ['envUri']);
+    }
+
+    public function environmentsDataset()
```
Comment (pnal) [inline test/unit/model/publishing/delivery/PublishingClassDeliveryServiceTest.php:198]: Are you using it?

File: test/unit/model/publishing/delivery/PublishingClassDeliveryServiceTest.php [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+
+        $results = $this->subject->publish($this->classMock, ['envUri']);
+        $this->assertInstanceOf(CallbackTaskInterface::class, reset($results));
+    }
+
+
+    /**
+     * @dataProvider environmentsDataset
+     */
+    public function testWrongEnvironments($existReturn)
```
Comment (pnal) [inline test/unit/model/publishing/delivery/PublishingClassDeliveryServiceTest.php:136]: unused parameter

**Review by pnal — APPROVED** [review]
Great job! Just a few comments
- [x]  New code is covered by tests (if applicable)
- [x]  Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ]  New code is respecting code style rules
- [ ]  New code is respecting best practices
- [x]  New code is not subject to concurrency issues (if applicable)
- [x]  Feature is working correctly on my local machine (if applicable)
- [x]  Acceptance criteria are respected
- [x]  Pull request title and description are meaningful

**Comment by AntonNeviarouski** [comment]
QA verified on http://test-bart.playground.kitchen.it.taocloud.org:49191/

**Pull request merged**